### PR TITLE
LPD-22301, LPD-22302 UndoRedo Component, Handling of metaFields and data-engine fields

### PR DESCRIPTION
**Pull request by NemethNorbert**
This PR is also pending on review from the forms team: https://github.com/liferay-forms/liferay-portal/pull/2538
and from the frontend-infra team: https://github.com/liferay-frontend/liferay-portal/pull/4223

# What is included?
LPD-22301](https://liferay.atlassian.net/browse/LPD-22301) , [LPD-22302](https://liferay.atlassian.net/browse/LPD-22302)
This is the first PR that includes the basic functionality for the epics.

1. Creating the FeatureFlag LPD-22301
2. Creating the UndoRedo component, responsible for storing and restoring the state of metadataFields and firing the undo redo events
3. Creating the HistoryReducer in data-engine, responsible for saving and restoring a snapshot of the state and firing the autoSave event onblur when a change is registered
4. Modifying input_localized to fire the autoSave event onblur, when a change is registered
5. A small change to make sure the rich-text field updated itself when the state is restroed within data-engine, I included this change so we can test it with a basic webcontent

# How to test?
Include feature flags:
```
feature.flag.LPD-11228=true
feature.flag.LPS-141392=true
feature.flag.LPD-15596=true
feature.flag.LPD-22301=true
```

1. Create a basic web content
2. Fill the title
3. Fill the content
4. Fill the description Field
5. Change Locale
6. Modify the title, content than the description Field
7. Use Undo and Redo buttons to test the functionality

# What is missing?

1. Default Locale change is not handled
2. There are DDM fields that do not update when the state is restored
3. Undo Redo button is not disabled while the state is restored

# Tests
There are no included tests, as this is just the first PR for the stories

### CODE REVIEW

**Comment by jkappler** [comment]
Removing backend review label since the changes are mostly frontend

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+	return (
+		<div>
+			<ClayButtonWithIcon
+				aria-label={Liferay.Language.get('undo')}
+				className="btn-monospaced"
+				disabled={step <= 0}
+				displayType="secondary"
+				onClick={() => {
+					Liferay.fire('undo');
```
Comment (ambrinchaudhary) [inline modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js:191]: We are not listening these events in other components, so no need to fire this (nor the 'redo' event)

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+		Liferay.on('autoSave', handleAutoSave);
+
+		return () => {
+			Liferay.detach('autoSave', handleAutoSave);
+		};
+	}, [handleAutoSave]);
+
+	useEffect(() => {
+		setTimeout(() => {
+			Liferay.fire('autoSave', {fieldName: 'Reset'});
```
Comment (ambrinchaudhary) [inline modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js:178]: This event 'autoSave' confuses me with the other functionality for AutoSave as draft for Web Content (which I don't know if is triggered also every second, but are different things). 

I would distinguish between both. We can call this event something like "storeState", or something similar. What do you think?

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+export default function UndoRedo({
+	initialDefaultLanguageId,
+	languageId,
+	namespace,
+}) {
+	const [
+		// eslint-disable-next-line no-unused-vars
+		{defaultLanguageId, history, selectedLanguageId, step},
+		setState,
+	] = useState({
```
Comment (ambrinchaudhary) [inline modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js:24]: Let's use the useState pattern for this `[x, setX] = useState({..}) `
I'm not sure it the name "state" is the appropiate, but right now I have no better suggestions so let's keep it as 
`const [state, setState] = useState({...})`

**Review by ambrinchaudhary — CHANGES_REQUESTED** [review]
Hey @NemethNorbert, 
thank you very much for sending this PR. It is huge and there are lot of things implemented, and although I still need some more time to check if we can reorganize better the` UndoRedo.js` components (maybe divide in subcomponents like it is done in PageEditor), I would like to request some changes meanwhile. 

The thing that most worries me are Feature Flags. We put FF only under Epics, not for stories. So only the Feature Flag `LPD-11228` should be necessary. Replace old `LPS-141392` with `LPD-11228`, remove the new `LPD-22301` FF created, and the other LPD-15596 that you mention in the pull description is not necessary for this. 

As said, I will continue on Monday, and I am also checking modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/undo/Undo.js to see if we can do something similar.

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js (hunk 1 above)
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js]: We are listening in data-engine. Restoring data-engine to a previous or future state requires of firing this event

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js (hunk 3 above)
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js:24]: I would rather keep the state destructured as it is more readable as well as easier to access the fields.

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js (hunk 2 above)
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js]: I renamed the event

**Comment by NemethNorbert** [comment]
Hey @ambrinchaudhary ,

I force pushed the requested changes, as well as I prefixed the events, that was requested from tha FEI team

**Comment by DiegoHu97** [comment]
Resent here: https://github.com/liferay-content-management/liferay-portal/pull/5559

**Pull request closed without merging**